### PR TITLE
NA - Add the ability to define custom errorLogger in Classy Node.

### PR DESCRIPTION
**Pull request by ESoch**
Add errorLogger param in so that we can manage errors that occur in requests OR in the ClassyNode instance.

### CODE REVIEW

File: src/ClassyResource/_makeRequest.js [hunk 1]
```diff
@@ -11,70 +11,109 @@ import request from 'request';
...
+            const errorString = body ? JSON.stringify(body) : 'Non-200 response';
+
+            error = new Error(errorString);
+            error.statusCode = response ? response.statusCode : undefined;
+          }
+
+          /**
+           * Pass the error to _errorLogger if defined
+           */
+          if (this._errorLogger) {
```
Comment (jgrisafe) [inline src/ClassyResource/_makeRequest.js]: I wonder if we should be logging all errors or maybe exclude 403s and 404s?

Comment (ESoch) [reply src/ClassyResource/_makeRequest.js]: Hmm, not sure. I feel like that would maybe be up to the client (implementing errorLogger) to decide.

Comment (jgrisafe) [reply src/ClassyResource/_makeRequest.js]: good point good point

File: README.md [hunk 2]
```diff
@@ -61,6 +61,35 @@ Each resource can contain several basic methods (retrieve, update, delete). Each
...
+    apiKey: process.env.BUGSNAG_API_KEY
+  });
+
+  errorLogger = (error, other) => bugsnag.notify(error, { metaData: other });
+}
+
+const classy = new Classy({
+  clientId: process.env.CLIENT_ID,
+  clientSecret: process.env.CLIENT_SECRET,
+  baseUrl: 'https://stagingapi.stayclassy.org',
```
Comment (mlingner) [inline README.md]: Any concern with exposing the staging URL in a public repo?

Comment (ESoch) [reply README.md]: I don't think it should make a difference but worth changing regardless.

Comment (jgrisafe) [reply README.md]: Hello friend @mlingner, thanks for pitching in. Yeah Ethan get it outta there!

Comment (mlingner) [reply README.md]: Big open source and Classy fan 😃

File: src/ClassyResource/_makeRequest.js [hunk 3]
```diff
@@ -23,58 +23,103 @@ export default function _makeRequest(path, method, headers, form, data) {
...
       requestParams.body = JSON.stringify(data);
+    } catch (e) {
+      /**
+       * Pass the error to _errorLogger if defined
+       */
+      if (this._errorLogger) {
+        this._errorLogger(e, {
+          data,
+          requestParams
+        });
```
Comment (jgrisafe) [inline src/ClassyResource/_makeRequest.js]: should this rethrow the error?

Comment (ESoch) [reply src/ClassyResource/_makeRequest.js]: Ah yeah, good catch.

File: src/ClassyResource/_refreshAppToken.js [hunk 4]
```diff
@@ -4,18 +4,43 @@
...
  */
 export default function _refreshAppToken() {
-  const promise = new Promise((resolve, reject) => {
-    if (!this._classy.appToken) {
-      this._classy.getAppToken().then((response) => {
-        this._classy.setAppToken(response);
-        resolve(response);
-      }, (error) => {
-        reject(error);
+  try {
```
Comment (jgrisafe) [inline src/ClassyResource/_refreshAppToken.js]: this one seems redundant because the promise will handle any errors that occur inside it

File: src/ClassyResource/main.js [hunk 5]
```diff
@@ -24,6 +24,11 @@ export default class ClassyResource {
       });
     }
 
+    /** Set up Error Logger (if defined) */
+    if (this._classy.errorLogger) {
```
Comment (jgrisafe) [inline src/ClassyResource/main.js]: why is this needed?

Comment (ESoch) [reply src/ClassyResource/main.js]: Honestly, the ClassyResource vs. Classy convention in Classy-Node is odd and I don't really understand the intent. I did it this way to maintain the existing pattern. It seems like the intent originally was to define Classy defaults using the Classy content and use the ClassyResource to override those defaults if need be.

File: src/ClassyResource/_makeRequest.js [hunk 6]
```diff
@@ -23,58 +23,105 @@ export default function _makeRequest(path, method, headers, form, data) {
...
+    headers: headers,
+    rejectUnauthorized: false,
+    form: form,
+    encoding: encoding
+  };
+
+  if (method === 'GET') {
+    requestParams.qs = data;
+  } else {
+    try {
```
Comment (jgrisafe) [inline src/ClassyResource/_makeRequest.js]: had side conversation - stringify isn't as finicky as parse
